server/src/main/java/org/elasticsearch/action/admin/indices/datastream/DataStreamsStatsAction.java

```diff
@@ -0,0 +1,377 @@
+            return state.blocks().globalBlockedException(ClusterBlockLevel.METADATA_READ);
+        }
+
+        @Override
+        protected ClusterBlockException checkRequestBlock(ClusterState state, Request request, String[] concreteIndices) {
+            return state.blocks().indicesBlockedException(ClusterBlockLevel.METADATA_READ, concreteIndices);
+        }
+
+        @Override
+        protected ShardsIterator shards(ClusterState clusterState, Request request, String[] concreteIndices) {
+            String[] requestIndices = request.indices();
+            if (requestIndices == null || requestIndices.length == 0) {
+                requestIndices = new String[]{"*"};
+            }
+            List<String> abstractionNames = indexAbstractionResolver.resolveIndexAbstractions(requestIndices, request.indicesOptions(),
+                clusterState.getMetadata());
+            SortedMap<String, IndexAbstraction> indicesLookup = clusterState.getMetadata().getIndicesLookup();
+
+            String[] concreteDatastreamIndices = abstractionNames.stream().flatMap(abstractionName -> {
+                IndexAbstraction indexAbstraction = indicesLookup.get(abstractionName);
+                assert indexAbstraction != null;
+                if (indexAbstraction.getType() == IndexAbstraction.Type.DATA_STREAM) {
+                    IndexAbstraction.DataStream dataStream = (IndexAbstraction.DataStream) indexAbstraction;
+                    List<IndexMetadata> indices = dataStream.getIndices();
+                    return indices.stream().map(idx -> idx.getIndex().getName());
+                } else {
+                    return Stream.empty();
+                }
+            }).toArray(String[]::new);
+            return clusterState.getRoutingTable().allShards(concreteDatastreamIndices);
+        }
+
+        @Override
+        protected DataStreamShardStats shardOperation(Request request, ShardRouting shardRouting) throws IOException {
+            IndexService indexService = indicesService.indexServiceSafe(shardRouting.shardId().getIndex());
+            IndexShard indexShard = indexService.getShard(shardRouting.shardId().id());
+            // if we don't have the routing entry yet, we need it stats wise, we treat it as if the shard is not ready yet
+            if (indexShard.routingEntry() == null) {
+                throw new ShardNotFoundException(indexShard.shardId());
+            }
+            StoreStats storeStats = indexShard.storeStats();
+            IndexAbstraction indexAbstraction = clusterService.state().getMetadata().getIndicesLookup().get(shardRouting.getIndexName());
+            assert indexAbstraction != null;
+            IndexAbstraction.DataStream dataStream = indexAbstraction.getParentDataStream();
+            assert dataStream != null;
+            long maxTimestamp = 0L;
+            try (Engine.Searcher searcher = indexShard.acquireSearcher("data_stream_stats")) {
```

Comment:
I think we only need to capture this information from the latest backing index?
All other backing indices should have max timestamp that is lower than 
the timestamp in the latest index.

Maybe we need to collect the max timestamp from the two latest backing indices,
in the case the write index just rolled over.

Reply:
I'm not well versed in the costs for different Lucene operations. Is this operation heavy enough that we would see a benefit from only targeting the latest indices? If so, the most recent index is easy enough to discern because it's configured in the data stream to accept writes. Are the backing indices to be considered always going to be the same naming format (using numbers?), and do we think that's a stable enough practice to base stat collection logic on?

Reply:
This isn't a very expensive operations compared to other Lucene operations. However I think that for cold indices, not heaving to do this operation on most shards is a win.

> Are the backing indices to be considered always going to be the same naming format (using numbers?), and do we think that's a stable enough practice to base stat collection logic on?

For now it is stable, but when existing indices can be imported into a data stream not. Perhaps this should be based on the ordering of the indices in the data stream? The last index should always be the write index.

Reply:
I think for now it is ok to go to all shards and retrieve the max timestamp. This is an optimization that can be done later.